mlir/lib/Dialect/MIOpen/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -128,22 +148,166 @@ void affixThreadwiseCopyV2Attributes(ThreadwiseCopyV2Op top,
+      // spreads out the loaded elements to appropriate indices. If the
+      // vectorization dimension is the fastest-moving dimension of the loop, we
+      // don't need to do this
+      SmallVector<int64_t, 4> vectorIdxBounds(nUpper, 1);
+      vectorIdxBounds[vectorDim] = loadLength;
+      ArrayAttr loadedValIdxMap = makeLinearDomain(b, loc, vectorIdxBounds);
+
+      TransformingForOp scatterLoop = b.create<TransformingForOp>(
+          loc, ArrayRef<ValueRange>{linearInit, loop.getLowerCoords(1)},
+          ArrayRef<Attribute>{loadedValIdxMap, resultIdxMap}, vectorIdxBounds,
+          /*forceUnroll=*/true, /*useIndexDiffs=*/true, loopArg);
+
+      {
+        OpBuilder::InsertionGuard innerGuard(b);
+        b.setInsertionPointToStart(scatterLoop.getBody());
+        Value toScatter = b.create<vector::ExtractElementOp>(
+            loc, loaded, scatterLoop.getLowerCoords(0)[0]);
+        Value toYieldInner = b.create<vector::InsertElementOp>(
+            loc, toScatter, scatterLoop.getIterArgs()[0],
+            scatterLoop.getLowerCoords(1)[0]);
+        b.create<miopen::YieldOp>(loc, toYieldInner);
+      }
+      toYield = scatterLoop.getResults()[0];
+    } else {
+      toYield = b.create<InsertSliceOp>(loc, resultType, loaded, loopArg,
+                                        loop.getLowerCoords(1)[0]);
+    }
+  }
+  b.create<miopen::YieldOp>(loc, toYield);
+  return loop;
 }
 
-template <typename T, typename U>
-void affixBlockwiseCopyAttributes(T &bop, U &gop, OpBuilder &b, int vectorDim,
-                                  int blockwiseLoadLength,
-                                  int blockwiseStoreLength) {
-  bop->setAttr("block_size", gop->getAttr("block_size"));
-
-  bop->setAttr("source_vector_read_dim", b.getI32IntegerAttr(vectorDim));
-  bop->setAttr("dest_vector_write_dim", b.getI32IntegerAttr(vectorDim));
-  bop->setAttr("source_data_per_read",
-               b.getI32IntegerAttr(blockwiseLoadLength));
-  bop->setAttr("dest_data_per_write",
-               b.getI32IntegerAttr(blockwiseStoreLength));
+TransformingForOp createLdsStoreLoop(OpBuilder &b, Location loc, Value loaded,
+                                     Value buffer, ValueRange bufferStart,
+                                     Type storingType,
+                                     ArrayRef<int64_t> sliceLengths,
+                                     uint32_t vectorDim) {
+  Type loadedType = loaded.getType();
+  bool fullyScalar = !loadedType.isa<ShapedType>();
+
+  int64_t storeLength = 1;
+  if (auto storingVectorType = storingType.dyn_cast<VectorType>())
+    storeLength = storingVectorType.getNumElements();
+
+  size_t nUpper = bufferStart.size();
+  bool complexVectorStore = (storeLength > 1) && (vectorDim != nUpper - 1);
+
+  Value zero = b.createOrFold<ConstantIndexOp>(loc, 0);
+  SmallVector<Value, 5> linearInit(nUpper, zero);
+
+  ArrayAttr bufferTransforms;
+  buffer = untransform(b, buffer, bufferTransforms);
```

Comment:
Without a proper documentation or some examples for `untransform`, this line is hard to understand.

Reply:
Hopefully the new version is better?